drivers/gpio/gpio_stm32.c

```diff
@@ -376,6 +490,75 @@ static int gpio_stm32_read(struct device *dev, int access_op,
 	return 0;
 }
 
+static int gpio_stm32_pin_interrupt_configure(struct device *dev,
+		unsigned int pin, enum gpio_int_mode mode,
+		enum gpio_int_trig trig)
+{
+	const struct gpio_stm32_config *cfg = dev->config->config_info;
+	struct gpio_stm32_data *data = dev->driver_data;
+	int edge = 0;
+	int err = 0;
+
+#if defined(CONFIG_STM32H7_DUAL_CORE)
+	while (LL_HSEM_1StepLock(HSEM, LL_HSEM_ID_1)) {
+	}
+#endif /* CONFIG_STM32H7_DUAL_CORE */
+
+	if (mode == GPIO_INT_MODE_DISABLED) {
+		if (gpio_stm32_int_enabled_port(pin) == cfg->port) {
+			stm32_exti_disable(pin);
+			stm32_exti_unset_callback(pin);
+			stm32_exti_trigger(pin, STM32_EXTI_TRIG_NONE);
+		}
+
+		data->cb_pins &= ~BIT(pin);
+
+		goto release_lock;
+	}
+
+	/* Level trigger interrupts not supported */
+	if (mode == GPIO_INT_MODE_LEVEL) {
+		err = -ENOTSUP;
+		goto release_lock;
+	}
+
+	if (stm32_exti_set_callback(pin, cfg->port,
+				    gpio_stm32_isr, dev) != 0) {
+		err = -EBUSY;
+		goto release_lock;
+	}
+
+	data->cb_pins |= BIT(pin);
+
+	gpio_stm32_enable_int(cfg->port, pin);
+
+	switch (trig) {
+	case GPIO_INT_TRIG_LOW:
+		edge = STM32_EXTI_TRIG_FALLING;
+		break;
+	case GPIO_INT_TRIG_HIGH:
+		edge = STM32_EXTI_TRIG_RISING;
+		break;
+	case GPIO_INT_TRIG_BOTH:
+		edge = STM32_EXTI_TRIG_BOTH;
+		break;
+	}
+
+	stm32_exti_trigger(pin, edge);
+
+	if (stm32_exti_enable(pin) != 0) {
+		err = -EIO;
+		goto release_lock;
```

Comment:
In a meantime we've called a few exti functions that seem to acquire/use resources: `stm32_exti_set_callback`, `gpio_stm32_enable_int`, `stm32_exti_trigger`. Do we need to do some clean up here?

Reply:
Actually `stm32_exti_enable` should be void.